### PR TITLE
resource/aws_elasticsearch_domain DomainEndpointOptions

### LINKED ISSUE #10424: Support "Require https" on elasticsearch
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

AWS announced support for https required on elasticsearch domains, Would it be possible to implement?

https://aws.amazon.com/about-aws/whats-new/2019/10/amazon-elasticsearch-service-provides-option-to-mandate-https/?ck_subscriber_id=512831964

Add `require_https` argument to elasticsearch resources

* aws_elasticsearch

### Potential Terraform Configuration

```hcl
resource "aws_elasticsearch_domain" "example" {
  domain_name           = "example"
  elasticsearch_version = "1.5"
  require_https = true

  cluster_config {
    instance_type = "r4.large.elasticsearch"
  }

  snapshot_options {
    automated_snapshot_start_hour = 23
  }

  tags = {
    Domain = "TestDomain"
  }
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by barrytam20**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #10424

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_elasticsearch_domain: Add `require_https` argument
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

**Comment by au78** [comment]
@barrytam20 - do you have an update on when this will completed?

**Comment by barrytam20** [comment]
@au78 i think the code should be ready for review, im working on running the tests.

**Comment by barrytam20** [comment]
im getting the following error for a few of my tests, ill have to investigate whats going on
```
--- FAIL: TestAccAWSElasticSearchDomain_internetToVpcEndpoint (1464.00s)
    testing.go:615: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=1) {
         (string) (len=13) "require_https": (string) (len=5) "false"
        }
```

**Comment by barrytam20** [comment]
fixed those tests. my aws account has a limited amount of VPCs and thus it looks like the vpc based tests are failing. Would anyone be able to help me run those?

```
make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSElasticSearchDomain* -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSElasticSearchDomainPolicy_basic
=== PAUSE TestAccAWSElasticSearchDomainPolicy_basic
=== RUN   TestAccAWSElasticSearchDomain_basic
=== PAUSE TestAccAWSElasticSearchDomain_basic
=== RUN   TestAccAWSElasticSearchDomain_RequireHTTPS
=== PAUSE TestAccAWSElasticSearchDomain_RequireHTTPS
=== RUN   TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== PAUSE TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== RUN   TestAccAWSElasticSearchDomain_withDedicatedMaster
=== PAUSE TestAccAWSElasticSearchDomain_withDedicatedMaster
=== RUN   TestAccAWSElasticSearchDomain_duplicate
=== PAUSE TestAccAWSElasticSearchDomain_duplicate
=== RUN   TestAccAWSElasticSearchDomain_v23
=== PAUSE TestAccAWSElasticSearchDomain_v23
=== RUN   TestAccAWSElasticSearchDomain_complex
=== PAUSE TestAccAWSElasticSearchDomain_complex
=== RUN   TestAccAWSElasticSearchDomain_vpc
=== PAUSE TestAccAWSElasticSearchDomain_vpc
=== RUN   TestAccAWSElasticSearchDomain_vpc_update
=== PAUSE TestAccAWSElasticSearchDomain_vpc_update
=== RUN   TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== PAUSE TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== RUN   TestAccAWSElasticSearchDomain_LogPublishingOptions
=== PAUSE TestAccAWSElasticSearchDomain_LogPublishingOptions
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== PAUSE TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
=== PAUSE TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
=== RUN   TestAccAWSElasticSearchDomain_policy
=== PAUSE TestAccAWSElasticSearchDomain_policy
=== RUN   TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== PAUSE TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== RUN   TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== PAUSE TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== RUN   TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== PAUSE TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== RUN   TestAccAWSElasticSearchDomain_tags
=== PAUSE TestAccAWSElasticSearchDomain_tags
=== RUN   TestAccAWSElasticSearchDomain_update
=== PAUSE TestAccAWSElasticSearchDomain_update
=== RUN   TestAccAWSElasticSearchDomain_update_volume_type
=== PAUSE TestAccAWSElasticSearchDomain_update_volume_type
=== RUN   TestAccAWSElasticSearchDomain_update_version
=== PAUSE TestAccAWSElasticSearchDomain_update_version
=== CONT  TestAccAWSElasticSearchDomainPolicy_basic
=== CONT  TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== CONT  TestAccAWSElasticSearchDomain_basic
=== CONT  TestAccAWSElasticSearchDomain_tags
=== CONT  TestAccAWSElasticSearchDomain_LogPublishingOptions
=== CONT  TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== CONT  TestAccAWSElasticSearchDomain_vpc_update
=== CONT  TestAccAWSElasticSearchDomain_vpc
=== CONT  TestAccAWSElasticSearchDomain_complex
=== CONT  TestAccAWSElasticSearchDomain_v23
=== CONT  TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== CONT  TestAccAWSElasticSearchDomain_duplicate
=== CONT  TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== CONT  TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== CONT  TestAccAWSElasticSearchDomain_policy
=== CONT  TestAccAWSElasticSearchDomain_withDedicatedMaster
=== CONT  TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== CONT  TestAccAWSElasticSearchDomain_RequireHTTPS
=== CONT  TestAccAWSElasticSearchDomain_update_version
=== CONT  TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
--- FAIL: TestAccAWSElasticSearchDomain_vpc (13.35s)
    testing.go:615: Step 0 error: errors during apply:
        
        Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.
        	status code: 400, request id: ef9bd6a3-77f8-42ae-af6f-10501125d530
        
          on /var/folders/f2/3m1142tj4bx77wg0yptbv73m0000gn/T/tf-test847602425/main.tf line 6:
          (source code not available)
        
        
=== CONT  TestAccAWSElasticSearchDomain_update_volume_type
--- FAIL: TestAccAWSElasticSearchDomain_vpc_update (19.00s)
    testing.go:615: Step 0 error: errors during apply:
        
        Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.
        	status code: 400, request id: 64249d42-6f7e-4ba9-b290-7352317a2d5c
        
          on /var/folders/f2/3m1142tj4bx77wg0yptbv73m0000gn/T/tf-test643060975/main.tf line 6:
          (source code not available)
        
        
=== CONT  TestAccAWSElasticSearchDomain_update
--- PASS: TestAccAWSElasticSearchDomain_LogPublishingOptions (827.62s)
--- PASS: TestAccAWSElasticSearchDomain_RequireHTTPS (863.80s)
--- PASS: TestAccAWSElasticSearchDomain_duplicate (864.37s)
--- PASS: TestAccAWSElasticSearchDomain_NodeToNodeEncryption (894.85s)
--- PASS: TestAccAWSElasticSearchDomain_complex (1034.49s)
--- PASS: TestAccAWSElasticSearchDomain_tags (1087.19s)
--- PASS: TestAccAWSElasticSearchDomain_policy (1098.27s)
--- FAIL: TestAccAWSElasticSearchDomain_internetToVpcEndpoint (1180.14s)
    testing.go:615: Step 2 error: errors during apply:
        
        Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached.
        	status code: 400, request id: a2d3bb6b-9b11-4b0a-94c5-f8e77c3e0440
        
          on /var/folders/f2/3m1142tj4bx77wg0yptbv73m0000gn/T/tf-test278058065/main.tf line 6:
          (source code not available)
        
        
--- PASS: TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key (1261.63s)
--- PASS: TestAccAWSElasticSearchDomain_v23 (1261.73s)
--- PASS: TestAccAWSElasticSearchDomainPolicy_basic (1266.37s)
--- PASS: TestAccAWSElasticSearchDomain_basic (1317.15s)
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove (1630.10s)
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsUpdate (1642.98s)
--- PASS: TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key (1657.04s)
--- PASS: TestAccAWSElasticSearchDomain_update (1677.52s)
--- PASS: TestAccAWSElasticSearchDomain_withDedicatedMaster (3055.49s)
--- PASS: TestAccAWSElasticSearchDomain_update_volume_type (3091.54s)
```

**Comment by hExPY** [comment]
+1

**Comment by sunil-apty** [comment]
@barrytam20 : Raise a support ticket to increase the limit of VPC in your account with a reason.
And your VPC limit will get increased in no time.

**Comment by barrytam20** [comment]
thanks @sunil-apty. I ran the tests after my service limit increase and it looks like everything passed

```
make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSElasticSearchDomain* -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSElasticSearchDomainPolicy_basic
=== PAUSE TestAccAWSElasticSearchDomainPolicy_basic
=== RUN   TestAccAWSElasticSearchDomain_basic
=== PAUSE TestAccAWSElasticSearchDomain_basic
=== RUN   TestAccAWSElasticSearchDomain_RequireHTTPS
=== PAUSE TestAccAWSElasticSearchDomain_RequireHTTPS
=== RUN   TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== PAUSE TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== RUN   TestAccAWSElasticSearchDomain_withDedicatedMaster
=== PAUSE TestAccAWSElasticSearchDomain_withDedicatedMaster
=== RUN   TestAccAWSElasticSearchDomain_duplicate
=== PAUSE TestAccAWSElasticSearchDomain_duplicate
=== RUN   TestAccAWSElasticSearchDomain_v23
=== PAUSE TestAccAWSElasticSearchDomain_v23
=== RUN   TestAccAWSElasticSearchDomain_complex
=== PAUSE TestAccAWSElasticSearchDomain_complex
=== RUN   TestAccAWSElasticSearchDomain_vpc
=== PAUSE TestAccAWSElasticSearchDomain_vpc
=== RUN   TestAccAWSElasticSearchDomain_vpc_update
=== PAUSE TestAccAWSElasticSearchDomain_vpc_update
=== RUN   TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== PAUSE TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== RUN   TestAccAWSElasticSearchDomain_LogPublishingOptions
=== PAUSE TestAccAWSElasticSearchDomain_LogPublishingOptions
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== PAUSE TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== RUN   TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
=== PAUSE TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
=== RUN   TestAccAWSElasticSearchDomain_policy
=== PAUSE TestAccAWSElasticSearchDomain_policy
=== RUN   TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== PAUSE TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== RUN   TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== PAUSE TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== RUN   TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== PAUSE TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== RUN   TestAccAWSElasticSearchDomain_tags
=== PAUSE TestAccAWSElasticSearchDomain_tags
=== RUN   TestAccAWSElasticSearchDomain_update
=== PAUSE TestAccAWSElasticSearchDomain_update
=== RUN   TestAccAWSElasticSearchDomain_update_volume_type
=== PAUSE TestAccAWSElasticSearchDomain_update_volume_type
=== RUN   TestAccAWSElasticSearchDomain_update_version
=== PAUSE TestAccAWSElasticSearchDomain_update_version
=== CONT  TestAccAWSElasticSearchDomainPolicy_basic
=== CONT  TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove
=== CONT  TestAccAWSElasticSearchDomain_LogPublishingOptions
=== CONT  TestAccAWSElasticSearchDomain_update_version
=== CONT  TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key
=== CONT  TestAccAWSElasticSearchDomain_vpc_update
=== CONT  TestAccAWSElasticSearchDomain_v23
=== CONT  TestAccAWSElasticSearchDomain_vpc
=== CONT  TestAccAWSElasticSearchDomain_duplicate
=== CONT  TestAccAWSElasticSearchDomain_withDedicatedMaster
=== CONT  TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig
=== CONT  TestAccAWSElasticSearchDomain_RequireHTTPS
=== CONT  TestAccAWSElasticSearchDomain_internetToVpcEndpoint
=== CONT  TestAccAWSElasticSearchDomain_NodeToNodeEncryption
=== CONT  TestAccAWSElasticSearchDomain_update_volume_type
=== CONT  TestAccAWSElasticSearchDomain_complex
=== CONT  TestAccAWSElasticSearchDomain_update
=== CONT  TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key
=== CONT  TestAccAWSElasticSearchDomain_basic
=== CONT  TestAccAWSElasticSearchDomain_tags
--- PASS: TestAccAWSElasticSearchDomain_duplicate (501.37s)
=== CONT  TestAccAWSElasticSearchDomain_CognitoOptionsUpdate
--- PASS: TestAccAWSElasticSearchDomain_RequireHTTPS (775.64s)
=== CONT  TestAccAWSElasticSearchDomain_policy
--- PASS: TestAccAWSElasticSearchDomain_complex (776.19s)
--- PASS: TestAccAWSElasticSearchDomain_encrypt_at_rest_specify_key (816.75s)
--- PASS: TestAccAWSElasticSearchDomain_NodeToNodeEncryption (833.26s)
--- PASS: TestAccAWSElasticSearchDomain_encrypt_at_rest_default_key (896.10s)
--- PASS: TestAccAWSElasticSearchDomain_LogPublishingOptions (930.38s)
--- PASS: TestAccAWSElasticSearchDomain_tags (968.65s)
--- PASS: TestAccAWSElasticSearchDomain_v23 (974.58s)
--- PASS: TestAccAWSElasticSearchDomain_vpc (1353.71s)
--- PASS: TestAccAWSElasticSearchDomain_basic (1380.30s)
--- PASS: TestAccAWSElasticSearchDomainPolicy_basic (1489.71s)
--- PASS: TestAccAWSElasticSearchDomain_policy (723.86s)
--- PASS: TestAccAWSElasticSearchDomain_internetToVpcEndpoint (1782.21s)
--- PASS: TestAccAWSElasticSearchDomain_update (2034.83s)
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsCreateAndRemove (2217.19s)
--- PASS: TestAccAWSElasticSearchDomain_vpc_update (2259.00s)
--- PASS: TestAccAWSElasticSearchDomain_CognitoOptionsUpdate (1965.38s)
--- PASS: TestAccAWSElasticSearchDomain_withDedicatedMaster (3006.57s)
--- PASS: TestAccAWSElasticSearchDomain_update_version (3748.28s)
--- PASS: TestAccAWSElasticSearchDomain_update_volume_type (3931.05s)
--- PASS: TestAccAWSElasticSearchDomain_ClusterConfig_ZoneAwarenessConfig (4895.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	4896.525s
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap	0.087s [no tests to run]
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags	0.089s [no tests to run]
```

File: aws/resource_aws_elasticsearch_domain.go [hunk 1]
```diff
@@ -361,6 +365,12 @@ func resourceAwsElasticSearchDomainCreate(d *schema.ResourceData, meta interface
 		input.AdvancedOptions = stringMapToPointers(v.(map[string]interface{}))
 	}
 
+	if d.Get("require_https").(bool) {
```
Comment (venky999) [inline aws/resource_aws_elasticsearch_domain.go]: This should option to include `TLSSecurityPolicy` as well

**Comment by au78** [comment]
Hello @barrytam20 ... any updates on when this will be completed?

**Comment by barrytam20** [comment]
sorry kind of forgot about this with the holidays and all😅

i ended up including a block for `DomainEndpointOptions` which includes `EnforceHTTPS` and `TLSSecurityPolicy`.

<details>
  <summary>make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_RequireHTTPS'</summary>

```
make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_RequireHTTPS'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSElasticSearchDomain_RequireHTTPS -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSElasticSearchDomain_RequireHTTPS
=== PAUSE TestAccAWSElasticSearchDomain_RequireHTTPS
=== CONT  TestAccAWSElasticSearchDomain_RequireHTTPS
--- PASS: TestAccAWSElasticSearchDomain_RequireHTTPS (603.85s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       605.096s
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap      0.366s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags 0.294s [no tests to run]
```
</details>

**Comment by au78** [comment]
thank you @barrytam20 ... it would really help if you can get this released by next week, please.

**Comment by barrytam20** [comment]
@au78 that is really not up to me :) i can see that this PR is moving up the list when sorted by 👍 but i dont think any maintainers have reviewed this just yet

**Comment by au78** [comment]
np... see whatever best you can do to move it up the ladder... really waiting for this one so that we can start adopting it without much fuss... ;)

**Comment by khurdz** [comment]
Hi and best wishes :)
Any idea when this PR would be reviewed and merged please ?

**Comment by barrytam20** [comment]
@bflad is there anything i can do to help get this one reviewed?

**Comment by krismassey** [comment]
Hi all, any potential for this to get released? I could really use this feature at the moment

File: aws/resource_aws_elasticsearch_domain.go [hunk 2]
```diff
@@ -443,6 +462,26 @@ func resourceAwsElasticSearchDomainCreate(d *schema.ResourceData, meta interface
...
+			domainOptions := elasticsearch.DomainEndpointOptions{}
+			if requireHTTPS, ok := s["require_https"]; ok {
+				domainOptions.EnforceHTTPS = aws.Bool(requireHTTPS.(bool))
+			}
+			if tls, ok := s["tls_security_policy"]; ok {
+				domainOptions.TLSSecurityPolicy = aws.String(tls.(string))
+			}
+			input.SetDomainEndpointOptions(&domainOptions)
+		}
+	}
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain.go]: This type of logic can usually be simplified by creating an "expand" function (similar to `expandESCognitoOptions` below): 👍 

```go
	if v, ok := d.GetOk("domain_endpoint_options"); ok {
		input.DomainEndpointOptions = expandESDomainEndpointOptions(v.([]interface{}))
	}

// ...

func expandESDomainEndpointOptions(l []interface{}) *elasticsearch.DomainEndpointOptions {
	if len(l) == 0 || l[0] == nil {
		return nil
	}

	m := l[0].(map[string]interface{})
	domainEndpointOptions := &elasticsearch.DomainEndpointOptions{}

	if v, ok := m["enforce_https"].(bool); ok {
		domainEndpointOptions.EnforceHTTPS = aws.Bool(v)
	}

	if v, ok := m["tls_security_policy"].(string); ok {
		domainEndpointOptions.TLSSecurityPolicy = aws.String(v)
	}

	return domainEndpointOptions
}
```

File: aws/resource_aws_elasticsearch_domain.go [hunk 3]
```diff
@@ -642,6 +681,13 @@ func resourceAwsElasticSearchDomainRead(d *schema.ResourceData, meta interface{}
 		d.Set("log_publishing_options", m)
 	}
 
+	if ds.DomainEndpointOptions != nil {
+		m := make(map[string]interface{})
+		m["require_https"] = ds.DomainEndpointOptions.EnforceHTTPS
+		m["tls_security_policy"] = ds.DomainEndpointOptions.TLSSecurityPolicy
+		d.Set("domain_endpoint_options", m)
+	}
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain.go]: Similar but opposite concept here with "flatten" functions 👍 Also, when using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types).

```go
	if err := d.Set("domain_endpoint_options", flattenESDomainEndpointOptions(ds.DomainEndpointOptions)); err != nil {
		return fmt.Errorf("error setting domain_endpoint_options: %s", err)
	}

// ...

func flattenESDomainEndpointOptions(domainEndpointOptions *elasticsearch.DomainEndpointOptions) []interface{} {
	if domainEndpointOptions == nil {
		return nil
	}

	m := map[string]interface{}{
		"enforce_https":       aws.BoolValue(domainEndpointOptions.EnforceHTTPS),
		"tls_security_policy": aws.StringValue(domainEndpointOptions.TLSSecurityPolicy),
	}

	return []interface{}{m}
}
```

File: aws/resource_aws_elasticsearch_domain.go [hunk 4]
```diff
@@ -684,6 +730,28 @@ func resourceAwsElasticSearchDomainUpdate(d *schema.ResourceData, meta interface
 		input.AdvancedOptions = stringMapToPointers(d.Get("advanced_options").(map[string]interface{}))
 	}
 
+	if d.HasChange("require_https") {
+		input.SetDomainEndpointOptions(&elasticsearch.DomainEndpointOptions{
+			EnforceHTTPS: aws.Bool(d.Get("require_https").(bool)),
+		})
+	}
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain.go]: This looks like an older iteration of this feature request 😄

File: aws/resource_aws_elasticsearch_domain.go [hunk 5]
```diff
@@ -684,6 +730,28 @@ func resourceAwsElasticSearchDomainUpdate(d *schema.ResourceData, meta interface
...
+			domainOptions := elasticsearch.DomainEndpointOptions{}
+			if requireHTTPS, ok := s["require_https"]; ok {
+				domainOptions.EnforceHTTPS = aws.Bool(requireHTTPS.(bool))
+			}
+			if tls, ok := s["tls_security_policy"]; ok {
+				domainOptions.TLSSecurityPolicy = aws.String(tls.(string))
+			}
+			input.SetDomainEndpointOptions(&domainOptions)
+		}
+	}
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain.go]: We can re-use the above mentioned "expand" function here:

```go
	if d.HasChange("domain_endpoint_options") {
		input.DomainEndpointOptions = expandESDomainEndpointOptions(d.Get("domain_endpoint_options").([]interface{}))
	}
```

File: aws/resource_aws_elasticsearch_domain_test.go [hunk 6]
```diff
@@ -87,6 +87,29 @@ func TestAccAWSElasticSearchDomain_basic(t *testing.T) {
...
+		CheckDestroy: testAccCheckESDomainDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccESDomainConfig_RequireHTTPS(ri),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckESDomainExists("aws_elasticsearch_domain.example", &domain),
+					testAccCheckESDomainEndpointOptions(true, "Policy-Min-TLS-1-2-2019-07", &domain),
+					resource.TestCheckResourceAttr(
+						"aws_elasticsearch_domain.example", "elasticsearch_version", "1.5"),
+					resource.TestMatchResourceAttr("aws_elasticsearch_domain.example", "kibana_endpoint", regexp.MustCompile(".*es.amazonaws.com/_plugin/kibana/")),
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain_test.go]: Nit: These checks look irrelevant to the testing of domain endpoint options and can be removed. 👍

File: aws/resource_aws_elasticsearch_domain_test.go [hunk 7]
```diff
@@ -87,6 +87,29 @@ func TestAccAWSElasticSearchDomain_basic(t *testing.T) {
...
 
+func TestAccAWSElasticSearchDomain_RequireHTTPS(t *testing.T) {
+	var domain elasticsearch.ElasticsearchDomainStatus
+	ri := acctest.RandInt()
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckESDomainDestroy,
+		Steps: []resource.TestStep{
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain_test.go]: We should include two additional `TestStep` that verify that imports and in-place updates work as expected with this new functionality, e.g.

```go
		Steps: []resource.TestStep{
			{
				Config: testAccESDomainConfig_DomainEndpointOptions(ri, true, "Policy-Min-TLS-1-0-2019-07"),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckESDomainExists("aws_elasticsearch_domain.example", &domain),
					testAccCheckESDomainEndpointOptions(true, "Policy-Min-TLS-1-0-2019-07", &domain),
				),
			},
			{
				ResourceName:      "aws_elasticsearch_domain.example",
				ImportState:       true,
				ImportStateId:     resourceId,
				ImportStateVerify: true,
			},
			{
				Config: testAccESDomainConfig_DomainEndpointOptions(ri, true, "Policy-Min-TLS-1-2-2019-07"),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckESDomainExists("aws_elasticsearch_domain.example", &domain),
					testAccCheckESDomainEndpointOptions(true, "Policy-Min-TLS-1-2-2019-07", &domain),
				),
			},
		},

// ...

func testAccESDomainConfig_DomainEndpointOptions(randInt int, enforceHttps bool, tlsSecurityPolicy string) string {
	return fmt.Sprintf(`
resource "aws_elasticsearch_domain" "example" {
  domain_name = "tf-test-%[1]d"

  domain_endpoint_options {
    enforce_https       = %[2]t
    tls_security_policy = %[3]q
  }

  ebs_options {
    ebs_enabled = true
    volume_size = 10
  }
}
`, randInt, enforceHttps, tlsSecurityPolicy)
}
```

File: website/docs/r/elasticsearch_domain.html.markdown [hunk 8]
```diff
@@ -240,6 +241,11 @@ The following arguments are supported:
 * `enabled` - (Required) Whether to enable encryption at rest. If the `encrypt_at_rest` block is not provided then this defaults to `false`.
 * `kms_key_id` - (Optional) The KMS key id to encrypt the Elasticsearch domain with. If not specified then it defaults to using the `aws/es` service KMS key.
 
+**domain_endpoint_options** supports the following attributes:
+
+* `require_https` - (Required) wether or not to require HTTPS
```
Comment (bflad) [inline website/docs/r/elasticsearch_domain.html.markdown]: Typo 😎 

```suggestion
* `enforce_https` - (Required) Whether or not to require HTTPS
```

File: website/docs/r/elasticsearch_domain.html.markdown [hunk 9]
```diff
@@ -240,6 +241,11 @@ The following arguments are supported:
...
+* `tls_security_policy` - (Optional) the TLS security policy that needs to be applied to the HTTPS endpoint. Defaults to `Policy-Min-TLS-1-2-2019-07`
```
Comment (bflad) [inline website/docs/r/elasticsearch_domain.html.markdown]: We would prefer not to choose a default in this situation, but should still mention the valid values.

```suggestion
* `tls_security_policy` - (Optional) The name of the TLS security policy that needs to be applied to the HTTPS endpoint. Valid values:  `Policy-Min-TLS-1-0-2019-07` and `Policy-Min-TLS-1-2-2019-07`. Terraform will only perform drift detection if a configuration value is provided.
```

**Comment by barrytam20** [comment]
thanks @bflad for the comments!

I've implemented the changes you request, but im getting some weird error with my tests. I'll keep looking into it

<details>
  <summary>make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_RequireHTTPS'</summary>

```
make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_RequireHTTPS'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSElasticSearchDomain_RequireHTTPS -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSElasticSearchDomain_RequireHTTPS
=== PAUSE TestAccAWSElasticSearchDomain_RequireHTTPS
=== CONT  TestAccAWSElasticSearchDomain_RequireHTTPS
--- FAIL: TestAccAWSElasticSearchDomain_RequireHTTPS (657.61s)
    testing.go:635: Step 1 error: ValidationException: 1 validation error detected: Value 'aws_sns_sms_id' at 'domainName' failed to satisfy constraint: Member must satisfy regular expression pattern: [a-z][a-z0-9\-]+
```
</details?

**Comment by barrytam20** [comment]
thanks @bflad!

ive implemented the changes you requested, and re-ran the tests

``` 
make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_RequireHTTPS'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSElasticSearchDomain_RequireHTTPS -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSElasticSearchDomain_RequireHTTPS
=== PAUSE TestAccAWSElasticSearchDomain_RequireHTTPS
=== CONT  TestAccAWSElasticSearchDomain_RequireHTTPS
--- PASS: TestAccAWSElasticSearchDomain_RequireHTTPS (1166.52s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       1167.762s
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap      0.247s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags 0.134s [no tests to run
```

File: website/docs/r/elasticsearch_domain.html.markdown [hunk 10]
```diff
@@ -224,6 +224,7 @@ The following arguments are supported:
 * `vpc_options` - (Optional) VPC related options, see below. Adding or removing this configuration forces a new resource ([documentation](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html#es-vpc-limitations)).
 * `log_publishing_options` - (Optional) Options for publishing slow logs to CloudWatch Logs.
 * `elasticsearch_version` - (Optional) The version of Elasticsearch to deploy. Defaults to `1.5`
+* `enforce_httpsnt_options` - (Optional) Domain endpoint HTTP(S) related options. See below.
```
Comment (bflad) [inline website/docs/r/elasticsearch_domain.html.markdown]: Looks like this fix got transposed between here and line 246 -- will fix on merge. 👍
